### PR TITLE
Implement JVM_VirtualThread[Mount|Unmount|Start|End]

### PR DESCRIPTION
**Pull request by thallium**
- Move code of APIs of JDK 20 such as JVM_VirtualThreadMountBegin
  to helpers so they can be called by APIs of both JDK 20 and 21
- Hide frames between mount/unmount begin and mount/unmount end

Fixes https://github.com/eclipse-openj9/openj9/issues/16984

### CODE REVIEW

**Comment by thallium** [comment]
@fengxue-IS FYI

**Comment by thallium** [comment]
@fengxue-IS can you take a look?

File: runtime/j9vm/javanextvmi.cpp [hunk 1]
```diff
@@ -272,8 +273,24 @@ exitVThreadTransitionCritical(J9VMThread *currentThread, j9object_t vthread)
...
+	Assert_SC_true(IS_JAVA_LANG_VIRTUALTHREAD(currentThread, currentThread->threadObject));
+
+	if (hide) {
+		Assert_SC_true(J9_ARE_NO_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
+		currentThread->privateFlags |= J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES;
+	} else {
+		Assert_SC_true(J9_ARE_ALL_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
+		currentThread->privateFlags &= ~(UDATA)J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES;
+	}
+}
```
Comment (fengxue-IS) [inline runtime/j9vm/javanextvmi.cpp:288]: Keeping in mind of the possible FastJNI code for this, I would like to see this being added to VM_VMHelpers and called by `JVM_VirtualThreadHideFrames` directly. no reason to create this helper which cannot be shared.
Note the assert will have to stay inside the JVM* function since VMHelpers cannot assert

File: runtime/j9vm/javanextvmi.cpp [hunk 2]
```diff
@@ -302,11 +319,15 @@ JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
 
 	vmFuncs->internalExitVMToJNI(currentThread);
 
+#if JAVA_SPEC_VERSION >= 20
+	virtualThreadHideFrames(currentThread, threadObj, TRUE);
+#endif /* JAVA_SPEC_VERSION >= 20 */
+
```
Comment (fengxue-IS) [inline runtime/j9vm/javanextvmi.cpp:325]: these can be directly added inside the `JVM_VirtualThreadMount/Unmount` functions to avoid duplicates.

**Review by fengxue-IS — CHANGES_REQUESTED** [review]

File: runtime/j9vm/javanextvmi.cpp [hunk 3]
```diff
@@ -272,8 +273,24 @@ exitVThreadTransitionCritical(J9VMThread *currentThread, j9object_t vthread)
 	J9OBJECT_I64_STORE(currentThread, vthread, currentThread->javaVM->virtualThreadInspectorCountOffset, 0);
 }
 
-JNIEXPORT void JNICALL
-JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+#if JAVA_SPEC_VERSION >= 20
+static void
+virtualThreadHideFrames(J9VMThread *currentThread, j9object_t vthread, BOOLEAN hide) {
+	Assert_SC_true(IS_JAVA_LANG_VIRTUALTHREAD(currentThread, currentThread->threadObject));
```
Comment (babsingh) [inline runtime/j9vm/javanextvmi.cpp:279]: `J9VMThread->threadObject` will change between `virtualThread[M|Un]ountBegin` and `virtualThread[M|Un]ountEnd`. `J9VMThread->threadObject` won't always be a virtual thread.

Comment (fengxue-IS) [reply runtime/j9vm/javanextvmi.cpp:279]: Hide frames should always be called from a virtual thread object to notify VM of modification to its currents states which should be ignored. this behaviour doesn't affect the mount/unmount operation as `HideFrames` should only be called by mounted virtual thread.

Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp:279]: This assertion is triggered when the `jvmti/events/FramePop/framepop02` test (https://github.com/eclipse-openj9/openj9/issues/16346) is run with the changes from this PR. Without the assertion, the failing `framepop02` test passes with the changes from this PR. The logic needs a revision to avoid triggering this assertion.

Comment (fengxue-IS) [reply runtime/j9vm/javanextvmi.cpp:279]: Ok, i see what you mean, the assertion should be placed directly inside `JVM_notifyJvmtiHideFrames` instead of the helper.

Comment (fengxue-IS) [reply runtime/j9vm/javanextvmi.cpp:279]: @thallium can yo move the assertion such that only `JVM_VirtualThreadHideFrames` triggers that.

**Comment by thallium** [comment]
@fengxue-IS Your comments have been addressed.

File: runtime/j9vm/javanextvmi.cpp [hunk 4]
```diff
@@ -246,6 +246,7 @@ JVM_IsPreviewEnabled(void)
 	}
 	return isPreviewEnabled;
 }
+#endif /* JAVA_SPEC_VERSION >= 19 */
```
Comment (fengxue-IS) [inline runtime/j9vm/javanextvmi.cpp:249]: this is not needed,   `enterVThreadTransitionCritical` and `exitVThreadTransitionCritical` is part of Java19

File: runtime/j9vm/javanextvmi.cpp [hunk 5]
```diff
@@ -272,8 +273,9 @@ exitVThreadTransitionCritical(J9VMThread *currentThread, j9object_t vthread)
 	J9OBJECT_I64_STORE(currentThread, vthread, currentThread->javaVM->virtualThreadInspectorCountOffset, 0);
 }
 
-JNIEXPORT void JNICALL
-JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+#if JAVA_SPEC_VERSION >= 19
```
Comment (fengxue-IS) [inline runtime/j9vm/javanextvmi.cpp:276]: same as above, not needed as exter/exitCritical is part of Java 19

File: runtime/oti/VMHelpers.hpp [hunk 6]
```diff
@@ -2058,6 +2058,16 @@ class VM_VMHelpers
 		return oldState;
 	}
 
-};
+#if JAVA_SPEC_VERSION >= 20
+	static VMINLINE void
+	virtualThreadHideFrames(J9VMThread *currentThread, j9object_t vthread, bool hide) {
```
Comment (fengxue-IS) [inline runtime/oti/VMHelpers.hpp:2063]: since vthread object is not used, we don't need to pass it. Also if you use `jboolean` instead of `bool`, then you can directly pass the `hide` argument rather than explicit `true/false`.

**Review by fengxue-IS — COMMENTED** [review]
code looks good, minor format suggestions.
@thallium can you confirm the failure seen in https://github.com/eclipse-openj9/openj9/issues/16346 is fixed

**Comment by dipak-bagadiya** [comment]
@fengxue-IS @thallium 
"test/hotspot/jtreg/serviceability/jvmti/events/FramePop/framepop02." 
test case execution fails.
```
STDERR:
15:13:29.377 0x272200  j9scar.209    *   ** ASSERTION FAILED ** at /root/openj9-20/openj9-openjdk-jdk20/openj9/runtime/j9vm/javanextvmi.cpp:279: ((isSameOrSuperClassOf((((J9RAMClassRef*)(&((currentThread)->javaVM)->jclConstantPool[(30)]))->value), ((0 != (currentThread)->compressObjectReferences) ? (struct J9Clas
```

File: runtime/oti/VMHelpers.hpp (hunk 6 above)
Comment (thallium) [reply runtime/oti/VMHelpers.hpp:2063]: Does the order of calling hideframes and mountBegin/End matter? Can I move it outside of the `if` statement?

**Comment by thallium** [comment]
> @fengxue-IS @thallium "test/hotspot/jtreg/serviceability/jvmti/events/FramePop/framepop02." test case execution fails.
> 
> ```
> STDERR:
> 15:13:29.377 0x272200  j9scar.209    *   ** ASSERTION FAILED ** at /root/openj9-20/openj9-openjdk-jdk20/openj9/runtime/j9vm/javanextvmi.cpp:279: ((isSameOrSuperClassOf((((J9RAMClassRef*)(&((currentThread)->javaVM)->jclConstantPool[(30)]))->value), ((0 != (currentThread)->compressObjectReferences) ? (struct J9Clas
> ```

@dipak-bagadiya Just want to confirm is this my latest change? I've moved the assertion so it won't be checked during mount/unmount.

File: runtime/redirector/forwarders.m4 [hunk 7]
```diff
@@ -415,5 +415,9 @@ _IF([JAVA_SPEC_VERSION >= 21],
 	[_X(JVM_VirtualThreadMount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide, jboolean firstMount)])
 _IF([JAVA_SPEC_VERSION >= 21],
 	[_X(JVM_VirtualThreadUnmount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide, jboolean lastUnmount)])
```
Comment (keithc-ca) [inline runtime/redirector/forwarders.m4:417]: The signatures of mount/unmount will have fewer parameters in Java 21.

File: runtime/j9vm/exports.cmake [hunk 8]
```diff
@@ -435,6 +435,8 @@ else()
 		JVM_IsForeignLinkerSupported
 		JVM_VirtualThreadMount
 		JVM_VirtualThreadUnmount
+		JVM_VirtualThreadStart
+		JVM_VirtualThreadEnd
```
Comment (keithc-ca) [inline runtime/j9vm/exports.cmake:439]: Please keep the list in alphabetical order.

File: runtime/j9vm/j9vmnatives.xml [hunk 9]
```diff
@@ -442,5 +442,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
 		<export name="JVM_IsForeignLinkerSupported"/>
 		<export name="JVM_VirtualThreadMount"/>
 		<export name="JVM_VirtualThreadUnmount"/>
+		<export name="JVM_VirtualThreadStart"/>
+		<export name="JVM_VirtualThreadEnd"/>
```
Comment (keithc-ca) [inline runtime/j9vm/j9vmnatives.xml:446]: Please keep the list in alphabetical order.

File: runtime/j9vm/javanextvmi.cpp [hunk 10]
```diff
@@ -272,14 +273,27 @@ exitVThreadTransitionCritical(J9VMThread *currentThread, j9object_t vthread)
 	J9OBJECT_I64_STORE(currentThread, vthread, currentThread->javaVM->virtualThreadInspectorCountOffset, 0);
 }
 
-JNIEXPORT void JNICALL
-JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+static void unsetParentVthread(J9VMThread *currentThread, jobject thread)
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:276]: Missing newline after `void`.

File: runtime/j9vm/javanextvmi.cpp [hunk 11]
```diff
@@ -444,10 +451,47 @@ JVM_VirtualThreadUnmountEnd(JNIEnv *env, jobject thread, jboolean lastUnmount)
 
 	vmFuncs->internalExitVMToJNI(currentThread);
 
-	Trc_SC_VirtualThreadUnmountEnd_Exit(currentThread, thread, lastUnmount);
+	Trc_SC_VirtualThreadUnmountEnd_Exit(currentThread, thread);
 }
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
+#if JAVA_SPEC_VERSION >= 19 && JAVA_SPEC_VERSION < 21
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:458]: Missing parentheses, suggest using "between" pattern:
```c
#if (19 <= JAVA_SPEC_VERSION) && (JAVA_SPEC_VERSION < 21)
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by babsingh** [comment]
@thallium Can you confirm if all of @keithc-ca's feedback has been addressed?

**Comment by thallium** [comment]
@babsingh Yes.

**Comment by babsingh** [comment]
@keithc-ca Planning to launch builds to test functionality. Can you please confirm if your review is complete?

File: runtime/j9vm/j9scar.tdf [hunk 12]
```diff
@@ -351,17 +351,17 @@ TraceEvent=Trc_SC_LoadLibrary_OpenShared_Decorate NoEnv Overhead=1 Level=3 Templ
...
 
-TraceEntry=Trc_SC_VirtualThreadUnmountBegin_Entry Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
-TraceExit=Trc_SC_VirtualThreadUnmountBegin_Exit Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
+TraceEntry=Trc_SC_VirtualThreadUnmountBegin_Entry Obsolete Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
+TraceExit=Trc_SC_VirtualThreadUnmountBegin_Exit Obsolete Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
 
-TraceEntry=Trc_SC_VirtualThreadUnmountEnd_Entry Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
-TraceExit=Trc_SC_VirtualThreadUnmountEnd_Exit Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
+TraceEntry=Trc_SC_VirtualThreadUnmountEnd_Entry Obsolete Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
+TraceExit=Trc_SC_VirtualThreadUnmountEnd_Exit Obsolete Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
```
Comment (keithc-ca) [inline runtime/j9vm/j9scar.tdf:365]: I think we need both the old and new tracepoints (the old ones should be used for Java 20).

File: runtime/j9vm/javanextvmi.cpp [hunk 13]
```diff
@@ -444,10 +452,47 @@ JVM_VirtualThreadUnmountEnd(JNIEnv *env, jobject thread, jboolean lastUnmount)
...
-	Trc_SC_VirtualThreadUnmountEnd_Exit(currentThread, thread, lastUnmount);
+	Trc_SC_VirtualThreadUnmountEnd_Exit(currentThread, thread);
 }
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
+#if (JAVA_SPEC_VERSION >= 19) && (JAVA_SPEC_VERSION < 21)
+JNIEXPORT void JNICALL
+JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+{
+	virtualThreadMountBegin(env, thread);
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:463]: Why does this ignore `firstMount`?

File: runtime/j9vm/javanextvmi.cpp [hunk 14]
```diff
@@ -444,10 +452,47 @@ JVM_VirtualThreadUnmountEnd(JNIEnv *env, jobject thread, jboolean lastUnmount)
...
+		TRIGGER_J9HOOK_VM_VIRTUAL_THREAD_END(vm->hookInterface, currentThread);
+		unsetParentVthread((J9VMThread *)env, thread);
+	}
+	virtualThreadUnmountBegin(env, thread);
+}
+
+JNIEXPORT void JNICALL
+JVM_VirtualThreadUnmountEnd(JNIEnv *env, jobject thread, jboolean lastUnmount)
+{
+	virtualThreadUnmountEnd(env, thread);
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:492]: Why does this ignore `lastUnmount`?

File: runtime/j9vm/javanextvmi.cpp [hunk 15]
```diff
@@ -512,6 +556,25 @@ JVM_IsForeignLinkerSupported()
 {
 	return JNI_TRUE;
 }
+
+JNIEXPORT void JNICALL
+JVM_VirtualThreadStart(JNIEnv* env, jobject vthread)
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:561]: Space before, not after `*`; likewise for `JVM_VirtualThreadEnd()`.

File: runtime/redirector/forwarders.m4 [hunk 16]
```diff
@@ -412,8 +412,12 @@ _IF([JAVA_SPEC_VERSION >= 20],
 _IF([JAVA_SPEC_VERSION >= 21],
 	[_X(JVM_IsForeignLinkerSupported, JNICALL, false, jboolean, void)])
 _IF([JAVA_SPEC_VERSION >= 21],
-	[_X(JVM_VirtualThreadMount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide, jboolean firstMount)])
+	[_X(JVM_VirtualThreadMount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide)])
 _IF([JAVA_SPEC_VERSION >= 21],
-	[_X(JVM_VirtualThreadUnmount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide, jboolean lastUnmount)])
```
Comment (keithc-ca) [inline runtime/redirector/forwarders.m4:417]: `JVM_VirtualThreadMount` and `JVM_VirtualThreadIUnmount` are still neede for Java 20 (with the old signatures).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/j9vm/javanextvmi.cpp (hunk 14 above)
Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp:492]: In our implementation, `lastUnmount` is handled completely in `JVM_VirtualThreadUnmountBegin`. So, it does not get used here. The same response applies to the question for `firstMount`: https://github.com/eclipse-openj9/openj9/pull/17125/files#r1220069964.

File: runtime/redirector/forwarders.m4 (hunk 16 above)
Comment (babsingh) [reply runtime/redirector/forwarders.m4:417]: These methods are introduced in JDK21. JDK20 uses the old methods JVM_VirtualThreadMountBegin|End and JVM_VirtualThreadUnmountBegin|End, which are included above. References:
- JDK20: https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/1e863069cbd09f7165b9834e64e3305a642c56e3/src/java.base/share/classes/java/lang/VirtualThread.java#L1066-L1076
- JDK21: https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/509718f021217fb1c1be512df54024e65b86b9fc/src/java.base/share/classes/java/lang/VirtualThread.java#L1093-L1099

File: runtime/j9vm/j9scar.tdf (hunk 12 above)
Comment (babsingh) [reply runtime/j9vm/j9scar.tdf:365]: - The new tracepoints are not needed.
- The old tracepoints can also be used in JDK21. In JDK21, we implicitly know that `JVM_VirtualThreadStart (firstMount=true)`, `JVM_VirtualThreadEnd (lastUnmount=true)`, `JVM_VirtualThreadMount(firstMount=false)` and `JVM_VirtualThreadUnmount (lastUnmount=false)`.
- In JDK20, the new tracepoints remove information about `firstMount` and `lastUnmount`. This information should be retained since it is valuable for debugging. We can pass `firstMount` and `lastUnmount` values to the new `virtualThread*` helper functions and keep using the old tracepoints.

Comment (babsingh) [reply runtime/j9vm/j9scar.tdf:365]: There are two commits in the PR. With the above feedback, we will revert some elements to the first commit. If we decide to go with this feedback, my suggestion will be to squash and summarize all changes in a single commit to avoid confusion.

File: runtime/j9vm/j9scar.tdf [hunk 17]
```diff
@@ -375,3 +375,15 @@ TraceExit=Trc_SC_UnloadLibrary_Exit NoEnv Overhead=1 Level=2 Template="JVM_Unloa
...
+TraceExit=Trc_SC_VirtualThreadMountBegin_Exit Overhead=1 Level=3 Template="thread = 0x%p"
+
+TraceEntry=Trc_SC_VirtualThreadMountEnd_Entry Overhead=1 Level=3 Template="thread = 0x%p"
+TraceExit=Trc_SC_VirtualThreadMountEnd_Exit Overhead=1 Level=3 Template="thread = 0x%p"
+
+TraceEntry=Trc_SC_VirtualThreadUnmountBegin_Entry Overhead=1 Level=3 Template="thread = 0x%p"
+TraceExit=Trc_SC_VirtualThreadUnmountBegin_Exit Overhead=1 Level=3 Template="thread = 0x%p"
+
+TraceEntry=Trc_SC_VirtualThreadUnmountEnd_Entry Overhead=1 Level=3 Template="thread = 0x%p"
+TraceExit=Trc_SC_VirtualThreadUnmountEnd_Exit Overhead=1 Level=3 Template="thread = 0x%p"
```
Comment (keithc-ca) [inline runtime/j9vm/j9scar.tdf:389]: The names of these tracepoints, and the names of the functions that use them should be updated to more closely correspond the the concepts they deal with in Java 21 (which no longer uses "Begin" or "End" suffixes).

File: runtime/j9vm/j9scar.tdf (hunk 12 above)
Comment (babsingh) [reply runtime/j9vm/j9scar.tdf:365]: After taking another look, the tracepoints should be added in the `JVM_*` functions, not in the `virtualThread*` helper functions, since the `_Entry` and `_Exit` tracepoints represent the entry and exit points of the `JVM_*` functions. This also implies that we should add new tracepoints: they should be named as per the new `JVM_*` functions (re: https://github.com/eclipse-openj9/openj9/pull/17125#discussion_r1220147949) and they should include all input parameters to the `JVM_*` functions.

File: runtime/redirector/forwarders.m4 [hunk 18]
```diff
@@ -412,8 +412,12 @@ _IF([JAVA_SPEC_VERSION >= 20],
...
+	[_X(JVM_VirtualThreadUnmount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide)])
+_IF([JAVA_SPEC_VERSION >= 21],
+	[_X(JVM_VirtualThreadStart, JNICALL, false, void, JNIEnv *env, jobject vthread)])
+_IF([JAVA_SPEC_VERSION >= 21],
+	[_X(JVM_VirtualThreadEnd, JNICALL, false, void, JNIEnv *env, jobject vthread)])
```
Comment (keithc-ca) [inline runtime/redirector/forwarders.m4:421]: Please insert additions so the group for Java 21 is sorted by name.

File: runtime/j9vm/javanextvmi.cpp [hunk 19]
```diff
@@ -474,36 +519,35 @@ JNIEXPORT void JNICALL
...
 JNIEXPORT void JNICALL
-JVM_VirtualThreadMount(JNIEnv* env, jobject vthread, jboolean hide, jboolean firstMount)
+JVM_VirtualThreadMount(JNIEnv* env, jobject vthread, jboolean hide)
 {
 	if (hide) {
-		JVM_VirtualThreadMountBegin(env, vthread, firstMount);
+		virtualThreadMountBegin(env, vthread);
 	} else {
-		JVM_VirtualThreadMountEnd(env, vthread, firstMount);
+		virtualThreadMountEnd(env, vthread);
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:540]: I think names using the suffixes "Begin" and "End" just serve to muddy the waters here; "hide" and "begin vs. end" would  (just based on the names) seem unrelated.

File: runtime/oti/VMHelpers.hpp [hunk 20]
```diff
@@ -2062,6 +2062,16 @@ class VM_VMHelpers
 		return oldState;
 	}
 
-};
+#if JAVA_SPEC_VERSION >= 20
+	static VMINLINE void
+	virtualThreadHideFrames(J9VMThread *currentThread, jboolean hide) {
```
Comment (keithc-ca) [inline runtime/oti/VMHelpers.hpp:2067]: Opening `{` belongs on the next line.

File: runtime/j9vm/j9scar.tdf (hunk 12 above)
Comment (thallium) [reply runtime/j9vm/j9scar.tdf]: @babsingh reverted to old tracepoints and added tracepoints for new functions.

File: runtime/oti/VMHelpers.hpp [hunk 21]
```diff
@@ -2062,6 +2062,17 @@ class VM_VMHelpers
...
+	virtualThreadHideFrames(J9VMThread *currentThread, jboolean hide)
+	{
+		if (hide) {
+			currentThread->privateFlags |= J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES;
+		} else {
+			currentThread->privateFlags &= ~(UDATA)J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES;
+		}
+	}
+#endif /* JAVA_SPEC_VERSION >= 20 */
+	};
```
Comment (keithc-ca) [inline runtime/oti/VMHelpers.hpp:2076]: This line should not change.

Comment (thallium) [reply runtime/oti/VMHelpers.hpp]: Fixed

File: runtime/j9vm/javanextvmi.cpp (hunk 11 above)
Comment (thallium) [reply runtime/j9vm/javanextvmi.cpp]: Corrected.

File: runtime/j9vm/javanextvmi.cpp [hunk 22]
```diff
@@ -443,11 +437,67 @@ JVM_VirtualThreadUnmountEnd(JNIEnv *env, jobject thread, jboolean lastUnmount)
...
+
+	Trc_SC_VirtualThreadMountEnd_Exit(currentThread, thread, firstMount);
+}
+
+JNIEXPORT void JNICALL
+JVM_VirtualThreadUnmountBegin(JNIEnv *env, jobject thread, jboolean lastUnmount)
+{
+	J9VMThread *currentThread = (J9VMThread *)env;
+
+	Trc_SC_VirtualThreadUnmountBegin_Entry(currentThread, thread);
```
[the remark this answers is not in the document]
Comment (thallium) [reply runtime/j9vm/javanextvmi.cpp]: Fixed

File: runtime/j9vm/javanextvmi.cpp (hunk 19 above)
Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp:549]: How should the code be changed for this feedback? begin and end represent the start and end of the virtual thread mount and unmount operations. hide toggles the switch to hide frames and disable certain JVMTI events between the begin and end operations.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by keithc-ca** [comment]
I merged this into #17514 and ran a small test program that used virtual threads:
```
19:30:32.718 0x30d900j9vmutil(j9jvmt.16     *   ** ASSERTION FAILED ** at /home/keithc/space/openj9/jdknext/openj9/runtime/util/eventframe.c:87: (!(currentThread->inNative))
```
Here's an excerpt of the native stack:
```
4XENATIVESTACK               raiseAssertion+0xf1 (0x00007FC1D7AC6261 [libj9trc29.so+0xe261])
4XENATIVESTACK               doTracePoint+0x965 (0x00007FC1D7AC9EC5 [libj9trc29.so+0x11ec5])
4XENATIVESTACK               javaTrace+0xa0 (0x00007FC1D7AC2010 [libj9trc29.so+0xa010])
4XENATIVESTACK               pushEventFrame+0x189 (0x00007FC1D692D519 [libj9jvmti29.so+0x35519])
4XENATIVESTACK               prepareForEvent+0xef (0x00007FC1D6913F3F [libj9jvmti29.so+0x1bf3f])
4XENATIVESTACK               jvmtiHookVirtualThreadStarted+0xab (0x00007FC1D691530B [libj9jvmti29.so+0x1d30b])
4XENATIVESTACK               J9HookDispatch+0x136 (0x00007FC1DC14F9C6 [libj9hookable29.so+0x19c6])
4XENATIVESTACK               JVM_VirtualThreadStart+0x9e (0x00007FC1DC40086E [libjvm.so+0x2086e])
4XENATIVESTACK               ffi_call_unix64+0x56 (0x00007FC1DC378D06 [libj9vm29.so+0x223d06])
4XENATIVESTACK               ffi_call_int+0x1bc (0x00007FC1DC377CBC [libj9vm29.so+0x222cbc])
4XENATIVESTACK               _ZN37VM_DebugBytecodeInterpreterCompressed3runEP10J9VMThread+0x1da73 (0x00007FC1DC260C63 [libj9vm29.so+0x10bc63])
4XENATIVESTACK               debugBytecodeLoopCompressed+0xd5 (0x00007FC1DC2431E5 [libj9vm29.so+0xee1e5])
4XENATIVESTACK                (0x00007FC1DC3068B2 [libj9vm29.so+0x1b18b2])
```

**Comment by thallium** [comment]
@keithc-ca The assertion failure should be fixed now by moving acquiring VM access from helpers to `JVM_*` functions.

File: runtime/j9vm/j9scar.tdf [hunk 23]
```diff
@@ -351,17 +351,17 @@ TraceEvent=Trc_SC_LoadLibrary_OpenShared_Decorate NoEnv Overhead=1 Level=3 Templ
...
+TraceExit=Trc_SC_VirtualThreadMountEnd_Exit Overhead=1 Level=3 Template="thread = %p, firstMount = %d"
 
-TraceEntry=Trc_SC_VirtualThreadUnmountBegin_Entry Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
-TraceExit=Trc_SC_VirtualThreadUnmountBegin_Exit Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
+TraceEntry=Trc_SC_VirtualThreadUnmountBegin_Entry Overhead=1 Level=3 Template="thread = %p, lastUnmount = %d"
+TraceExit=Trc_SC_VirtualThreadUnmountBegin_Exit Overhead=1 Level=3 Template="thread = %p, lastUnmount = %d"
 
-TraceEntry=Trc_SC_VirtualThreadUnmountEnd_Entry Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
-TraceExit=Trc_SC_VirtualThreadUnmountEnd_Exit Overhead=1 Level=3 Template="thread = 0x%p, lastUnmount = %d"
+TraceEntry=Trc_SC_VirtualThreadUnmountEnd_Entry Overhead=1 Level=3 Template="thread = %p, lastUnmount = %d"
```
Comment (keithc-ca) [inline runtime/j9vm/j9scar.tdf:363]: While you're here, please also correct other uses of `0x%p` in this file.

File: runtime/j9vm/javanextvmi.cpp [hunk 24]
```diff
@@ -272,16 +273,27 @@ exitVThreadTransitionCritical(J9VMThread *currentThread, j9object_t vthread)
...
-JNIEXPORT void JNICALL
-JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+static void
+unsetParentVthread(J9VMThread *currentThread, jobject thread)
+{
+	enterVThreadTransitionCritical(currentThread, thread);
+	/* Re-fetch reference as enterVThreadTransitionCritical may release VMAccess. */
+	j9object_t threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
+	j9object_t continuationObj = J9VMJAVALANGVIRTUALTHREAD_CONT(currentThread, threadObj);
+	/* Add reverse link from Continuation object to VirtualThread object, this let JVMTI code */
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:283]: Please finish each sentence with a period.

File: runtime/j9vm/javanextvmi.cpp [hunk 25]
```diff
@@ -346,26 +354,17 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
 	/* Allow thread to be inspected again. */
 	exitVThreadTransitionCritical(currentThread, threadObj);
 
-	if (firstMount) {
-		TRIGGER_J9HOOK_VM_VIRTUAL_THREAD_STARTED(vm->hookInterface, currentThread);
-	}
 	TRIGGER_J9HOOK_VM_VIRTUAL_THREAD_MOUNT(vm->hookInterface, currentThread);
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:358]: Please delete this blank line.

File: runtime/j9vm/javanextvmi.cpp [hunk 26]
```diff
@@ -474,44 +540,107 @@ JNIEXPORT void JNICALL
...
+{
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+
+	Trc_SC_VirtualThreadEnd_Entry(currentThread, vthread);
+
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+
+
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:635]: Just one blank line here, please.

File: runtime/j9vm/javanextvmi.cpp [hunk 27]
```diff
@@ -441,13 +428,91 @@ JVM_VirtualThreadUnmountEnd(JNIEnv *env, jobject thread, jboolean lastUnmount)
...
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
 
-	Trc_SC_VirtualThreadUnmountEnd_Exit(currentThread, thread, lastUnmount);
+	Trc_SC_VirtualThreadMountBegin_Entry(currentThread, thread, firstMount);
+
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+
+	virtualThreadMountBegin(env, thread);
+
+	vmFuncs->internalExitVMToJNI(vmThread);
```
[the remark this answers is not in the document]
Comment (thallium) [reply runtime/j9vm/javanextvmi.cpp]: Fixed.

File: runtime/j9vm/javanextvmi.cpp [hunk 28]
```diff
@@ -272,16 +273,27 @@ exitVThreadTransitionCritical(J9VMThread *currentThread, j9object_t vthread)
...
+	J9VMJDKINTERNALVMCONTINUATION_SET_VTHREAD(currentThread, continuationObj, NULL);
+	exitVThreadTransitionCritical(currentThread, threadObj);
+}
+
+/* Caller must have VMAccess. */
+static void
+virtualThreadMountBegin(JNIEnv *env, jobject thread)
 {
 	J9VMThread *currentThread = (J9VMThread *)env;
 	J9JavaVM *vm = currentThread->javaVM;
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:293]: `vm` is only used within the `if` statement below, this declaration should be moved there.

File: runtime/j9vm/javanextvmi.cpp [hunk 29]
```diff
@@ -315,9 +324,8 @@ JVM_VirtualThreadMountEnd(JNIEnv *env, jobject thread, jboolean firstMount)
 	j9object_t threadObj = NULL;
 	BOOLEAN runJ9Hooks = FALSE;
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:325]: `vmFuncs` is not used only once; suggest `vm->internalVMFunctions` be used there instead.
`rootVirtualThread` and `runJ9Hooks` are unused; please remove those declaractions.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by thallium** [comment]
@keithc-ca Your comments have been addressed.

**Comment by keithc-ca** [comment]
Any further comments, @babsingh or @fengxue-IS?

Builds in https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/608 have passed and I expect a similar result in https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/5 after which I am ready to merge #17514 (which includes this) together with https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/608 and https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/5.

**Review by babsingh — APPROVED** [review]
> Any further comments, @babsingh or @fengxue-IS?

No. LGTM.

**Review by fengxue-IS — APPROVED** [review]

**Comment by keithc-ca** [comment]
Merged via #17514.
